rust/xaynet-analytics/src/database/instance.rs

```
@@ -0,0 +1,116 @@
+        match IsarInstance::create(path, Self::MAX_SIZE, Self::get_schema()) {
+            Ok(instance) => Ok(Self { instance }),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn instance(&self) -> &IsarInstance {
+        &self.instance
+    }
+
+    pub fn get_all_as_bytes(
+        &self,
+        collection_name: &str,
+    ) -> Result<Vec<(&ObjectId, &[u8])>, IsarError> {
+        let collection = match self.get_collection(collection_name) {
+            Ok(collection) => collection,
+            Err(e) => return Err(e),
+        };
+        let transaction = match self.instance().begin_txn(false) {
+            Ok(transaction) => transaction,
+            Err(e) => return Err(e),
+        };
+        let _bytes = match self
+            .instance()
+            .create_query_builder(collection)
+            .build()
+            .find_all_vec(&transaction)
+        {
+            Ok(bytes) => bytes,
+            Err(e) => return Err(e),
+        };
+
+        // TODO: not sure how to proceed to parse [u8] using the collection schema. didn't find examples in Isar
+        unimplemented!()
+    }
+
+    pub fn put(&self, collection_name: &str, object: &[u8]) -> Result<(), IsarError> {
+        let transaction = match self.instance().begin_txn(false) {
+            Ok(transaction) => transaction,
+            Err(e) => return Err(e),
+        };
+        let collection = match self.get_collection(collection_name) {
+            Ok(collection) => collection,
+            Err(e) => return Err(e),
+        };
+        match collection.put(&transaction, None, object) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn get_object_builder(&self, collection_name: &str) -> Result<ObjectBuilder, IsarError> {
+        match self.get_collection(collection_name) {
+            Ok(collection) => Ok(collection.get_object_builder()),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn get_schema() -> Schema {
```

This should propagate the error.

---

done